PWGCF/MultiparticleCorrelations/Core/MuPa-MemberFunctions.h

```diff
@@ -1119,65 +1057,79 @@ void ResetEventByEventQuantities()
-      (tracks.size() > ceh_a.fEventCuts[eTotalMultiplicity][eMax])) {
+  // a) Event cuts on info available both in reconstructed and simulated data:
+  //   *) NumberOfEvents: => cut directly in void process( ... )
+  //   *) TotalMultiplicity:
+  if ((tracks->size() < ceh_a.fEventCuts[eTotalMultiplicity][eMin]) ||
+      (tracks->size() > ceh_a.fEventCuts[eTotalMultiplicity][eMax])) {
     return kFALSE;
   }
-
-  // SelectedTracks: => cut directly in void process( ... )
-
-  // Centrality: TBI
+  //   *) SelectedTracks: => cut directly in void process( ... )
+  //   *) Centrality: TBI
   //  if ((  TBI   < ceh_a.fEventCuts[eCentrality][eMin]) || (  TBI   >
   //  ceh_a.fEventCuts[eCentrality][eMax])) {
   //    return kFALSE;
   //  }
-
-  // Vertex_x:
-  if ((collision.posX() < ceh_a.fEventCuts[eVertex_x][eMin]) ||
-      (collision.posX() > ceh_a.fEventCuts[eVertex_x][eMax])) {
+  //   *) Vertex_x:
+  if ((collision->posX() < ceh_a.fEventCuts[eVertex_x][eMin]) ||
+      (collision->posX() > ceh_a.fEventCuts[eVertex_x][eMax])) {
     return kFALSE;
   }
-
-  // Vertex_y:
-  if ((collision.posY() < ceh_a.fEventCuts[eVertex_y][eMin]) ||
-      (collision.posY() > ceh_a.fEventCuts[eVertex_y][eMax])) {
+  //   *) Vertex_y:
+  if ((collision->posY() < ceh_a.fEventCuts[eVertex_y][eMin]) ||
+      (collision->posY() > ceh_a.fEventCuts[eVertex_y][eMax])) {
     return kFALSE;
   }
-
-  // Vertex_z:
-  if ((collision.posZ() < ceh_a.fEventCuts[eVertex_z][eMin]) ||
-      (collision.posZ() > ceh_a.fEventCuts[eVertex_z][eMax])) {
+  //   *) Vertex_z:
+  if ((collision->posZ() < ceh_a.fEventCuts[eVertex_z][eMin]) ||
+      (collision->posZ() > ceh_a.fEventCuts[eVertex_z][eMax])) {
     return kFALSE;
   }
 
-  // NContributors:
-  if ((collision.numContrib() < ceh_a.fEventCuts[eNContributors][eMin]) ||
-      (collision.numContrib() > ceh_a.fEventCuts[eNContributors][eMax])) {
-    return kFALSE;
-  }
+  // b) Event cuts on info available only in reconstructed data:
+  if (gProcessRec || gProcessRecSim) {
+    CollisionRec const* collisionRec = reinterpret_cast<CollisionRec const*>(collision);
+    // CollisionRec const& collision = *collisionRec;
+    //   *) NContributors:
+    if ((collisionRec->numContrib() < ceh_a.fEventCuts[eNContributors][eMin]) ||
+        (collisionRec->numContrib() > ceh_a.fEventCuts[eNContributors][eMax])) {
+      return kFALSE;
+    }
+    // ...
+  } // if (gProcessRec || gProcessRecSim) {
+
+  // c) Event cuts on info available only in simulated data:
+  if (gProcessSim) {
+    CollisionSim const* collisionSim = reinterpret_cast<CollisionSim const*>(collision);
+    // CollisionSim const& collision = *collisionSim;
+    //   *) Impact parameter:
+    if ((collisionSim->impactParameter() < ceh_a.fEventCuts[eImpactParameter][eMin]) ||
+        (collisionSim->impactParameter() > ceh_a.fEventCuts[eImpactParameter][eMax])) {
+      return kFALSE;
+    }
+    // ...
+  } // if (gProcessSim) {
```

Comment:
Have you checked that this actually executes correctly for an MC datafile at reco and gen levels?
To me you are creating a shortcut for the template mechanism which is based on execution code which, in my opinion, is quite dangerous. You are doing at execution time what the compiler can do at compiling time in a much more secure way

Reply:
Yes, i tested it successfully with MC datafile at reco and gen levels, but only locally. Still this has to be checked in Hyperloop.

I wonder why you are arguing that this is dangerous - do you have a concrete example at hand when this approach would fail?

I recalled now that I already used in previous PRs casting in the code, i.e. at the line:

`pw_a.fWeightsHist[ppe] = reinterpret_cast<TH1D*>(hist->Clone());`

The context is a bit different, though. But here for instance casting is mandatory, because Clone() returns a pointer to TObject (please see this [link](https://root.cern.ch/doc/master/classTH1.html#ae6c9d17415e3fa37a342f8b0c68602b5)). Would you argue that here also casting is dangerous?

Finally, MegaLinter doesn't complain to using reinterpret_cast.